File: sdk/demo/python/src/prohand_demo/test_hand.py
#!/usr/bin/env python3
"""
ProHand SDK FFI Demo: Test Hand - Individual Joint Testing

Tests each joint of each finger individually with cyclic motion.
"""

import sys
import argparse
import time
from .utils import DemoBase


class TestHandDemo(DemoBase):
    """Individual joint testing demo."""

    def __init__(self):
        super().__init__("ProHand Test Hand - Individual Joint Testing")

    def run(
        self,
        command_endpoint: str,
        status_endpoint: str,
        hand_streaming_endpoint: str,
        wrist_streaming_endpoint: str,
        delay: float,
        cycles: int,
    ) -> int:
        """Run the test hand demo."""
        self.banner()

        print("\nConnection parameters:")
        print(f"  Command endpoint:       {command_endpoint}")
        print(f"  Status endpoint:         {status_endpoint}")
        print(f"  Hand streaming endpoint: {hand_streaming_endpoint}")
        print(f"  Wrist streaming endpoint: {wrist_streaming_endpoint}")

        print("\nTest parameters:")
        print(f"  Delay between moves: {delay}s")
        print(f"  Cycles per joint:    {cycles}")

        client = None
        try:
            self.section("Connecting to IPC host with streaming...")
            client = self.sdk.ProHandClient(
                command_endpoint, status_endpoint, hand_streaming_endpoint, wrist_streaming_endpoint
            )
            self.success("Connected with streaming support!")

            # Verify connection with a ping
            self.section("Verifying connection...")
            client.send_ping()
            time.sleep(0.2)  # Wait for ping response
            self.success("Connection verified!")

            # Enable streaming mode on the driver
            self.section("Enabling streaming mode...")
            self.info("Telling driver to accept streaming commands...")
            client.set_streaming_mode(True)

            # Wait for streaming connection to be established (with verification)
            self.info("Waiting for streaming connection to be ready...")
            if not client.wait_for_streaming_ready(timeout=5.0):
                self.error("Streaming connection failed to establish!")
                self.info("This may happen if the driver is busy, device is not connected, or not in Running state.")
                return 1

            self.success("Streaming mode enabled! Commands will use streaming socket.")

            fingers = ["thumb", "index", "middle", "ring", "pinky"]
            joint_names = ["metacarpal", "proximal", "intermediate", "distal"]

            # Zero all fingers initially (use streaming mode)
            self.section("Zeroing all fingers...")
            zero_positions = [0.0] * 20  # 5 fingers × 4 joints
            zero_wrist_positions = [0.0, 0.0]  # 2 wrist joints
            client.send_hand_streams(zero_positions, 0.45)
            client.send_wrist_streams(zero_wrist_positions)
            time.sleep(1.0)

            # Test each joint of each finger
            # Joint layout: thumb[0-3], index[4-7], middle[8-11], ring[12-15], pinky[16-19]
            for finger_idx, finger in enumerate(fingers):
                for j in range(4):
                    joint_idx = finger_idx * 4 + j
                    self.section(f"{finger} - {joint_names[j]} (joint {joint_idx})")

                    # Determine range based on joint and finger
                    if j == 0 and finger != "thumb":
                        # Metacarpal (abduction)
                        min_deg, max_deg = -30.0, 30.0
                    else:
                        # Flexion joints
                        min_deg, max_deg = 0.0, 90.0

                    import math
                    min_rad = math.radians(min_deg)
                    max_rad = math.radians(max_deg)

                    # Run cycles (use streaming for high-frequency commands)
                    for cycle in range(cycles):
                        # Start with all zero (20 positions: 5 fingers × 4 joints)
                        positions = [0.0] * 20

                        # For distal joint (j==3), pre-flex intermediate
                        if j == 3 and finger != "thumb":
                            positions[joint_idx - 1] = math.radians(90.0)

                        # Move to max position (use streaming for high-frequency control)
                        positions[joint_idx] = max_rad
                        client.send_hand_streams(positions, 0.45)
                        client.send_wrist_streams(zero_wrist_positions)
                        time.sleep(delay)

                        # Move to min position
                        positions[joint_idx] = min_rad
                        client.send_hand_streams(positions, 0.45)
                        client.send_wrist_streams(zero_wrist_positions)
                        time.sleep(delay)

            # Return to zero (use streaming mode)
            self.section("Returning to zero position...")
            client.send_hand_streams(zero_positions, 0.45)
            client.send_wrist_streams(zero_wrist_positions)
            time.sleep(0.5)

            # Disable streaming mode
            self.section("Disabling streaming mode...")
            client.set_streaming_mode(False)
            time.sleep(0.2)

            self.success("Test hand demo completed!")
            return 0

        except self.sdk.ConnectionError as e:
            self.error(f"Connection failed: {e}")
            print("\nMake sure prohand-headless-ipc-host is running")
            return 1
        except self.sdk.ProHandError as e:
            self.error(f"Demo failed: {e}")
            return 1
        except KeyboardInterrupt:
            self.info("\nInterrupted by user")
            return 1
        except Exception as e:
            self.error(f"Unexpected error: {e}")
            import traceback
            traceback.print_exc()
            return 1
        finally:
            if client is not None:
                try:
                    client.set_streaming_mode(False)
                except Exception:
                    pass
                try:
                    client.close()
                except Exception:
                    pass


def main() -> int:
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="Test each joint of each finger individually",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--command-endpoint",
        type=str,
        default="ipc:///tmp/prohand-commands.ipc",
        help="ZMQ command endpoint"
    )
    parser.add_argument(
        "--status-endpoint",
        type=str,
        default="ipc:///tmp/prohand-status.ipc",
        help="ZMQ status endpoint"
    )
    parser.add_argument(
        "--hand-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-hand-streaming.ipc",
        help="ZMQ hand streaming endpoint",
    )
    parser.add_argument(
        "--wrist-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-wrist-streaming.ipc",
        help="ZMQ wrist streaming endpoint",
    )
    parser.add_argument(
        "--delay", type=float, default=0.2, help="Delay between movements (seconds)"
    )
    parser.add_argument(
        "--cycles",
        type=int,
        default=5,
        help="Number of cycles per joint"
    )
    args = parser.parse_args()

    demo = TestHandDemo()
    return demo.run(
        args.command_endpoint,
        args.status_endpoint,
        args.hand_streaming_endpoint,
        args.wrist_streaming_endpoint,
        args.delay,
        args.cycles,
    )


if __name__ == "__main__":
    sys.exit(main())

File: sdk/demo/cpp/src/test_hand.cpp
/**
 * ProHand SDK FFI Demo: Test Hand - Test Each Joint
 *
 * Tests each joint of each finger individually with cyclic motion.
 */

#include <iostream>
#include <cxxopts.hpp>
#include <chrono>
#include <thread>
#include <cmath>

#include <prohand_sdk/ProHandClient.hpp>
#include "prohand_demo/Utils.hpp"

using namespace prohand_sdk;
using namespace prohand_demo;  // For utils namespace

int main(int argc, char** argv) {
    cxxopts::Options options("test_hand", "Test each joint of each finger individually");

    options.add_options()("command-endpoint", "ZMQ command endpoint",
                          cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-commands.ipc"))("status-endpoint", "ZMQ status endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-status.ipc"))("hand-streaming-endpoint", "ZMQ hand streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-hand-streaming.ipc"))("wrist-streaming-endpoint", "ZMQ wrist streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-wrist-streaming.ipc"))("d,delay", "Delay between movements (seconds)",
                                                                                                                                                                                                                                                                            cxxopts::value<double>()->default_value("0.2"))("c,cycles", "Number of cycles per joint",
                                                                                                                                                                                                                                                                                                                            cxxopts::value<int>()->default_value("5"))("h,help", "Print usage");

    auto result = options.parse(argc, argv);
    
    if (result.count("help")) {
        std::cout << options.help() << std::endl;
        return 0;
    }

    auto commandEndpoint = result["command-endpoint"].as<std::string>();
    auto statusEndpoint = result["status-endpoint"].as<std::string>();
    auto handStreamingEndpoint = result["hand-streaming-endpoint"].as<std::string>();
    auto wristStreamingEndpoint = result["wrist-streaming-endpoint"].as<std::string>();
    auto delay = result["delay"].as<double>();
    auto cycles = result["cycles"].as<int>();

    utils::printBanner("ProHand Test Hand - Individual Joint Testing");

    std::cout << "\nTest parameters:\n";
    std::cout << "  Delay between moves: " << delay << "s\n";
    std::cout << "  Cycles per joint: " << cycles << "\n";
    std::cout << "  Command endpoint: " << commandEndpoint << "\n";
    std::cout << "  Status endpoint: " << statusEndpoint << "\n";

    try {
        utils::printSection("Connecting to IPC host with streaming...");
        ProHandClient client(commandEndpoint, statusEndpoint, handStreamingEndpoint, wristStreamingEndpoint);
        utils::printSuccess("Connected with streaming support!");

        // Verify connection with a ping
        utils::printSection("Verifying connection...");
        client.sendPing();
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        utils::printSuccess("Connection verified!");

        // Enable streaming mode on the driver
        utils::printSection("Enabling streaming mode...");
        utils::printInfo("Telling driver to accept streaming commands...");
        client.setStreamingMode(true);

        // Wait for streaming connection to be established (with verification)
        utils::printInfo("Waiting for streaming connection to be ready...");
        if (!client.waitForStreamingReady(5.0))
        {
            utils::printError("Streaming connection failed to establish!");
            utils::printInfo("This may happen if the driver is busy or under load.");
            utils::printInfo("Try running the demo again.");
            client.setStreamingMode(false);
            return 1;
        }

        utils::printSuccess("Streaming mode enabled! Commands will use streaming socket.");

        const std::vector<std::string> fingers = {"thumb", "index", "middle", "ring", "pinky"};
        const std::vector<std::string> jointNames = {
            "metacarpal", "proximal", "intermediate", "distal"};

        // Zero all fingers initially (use streaming mode)
        utils::printSection("Zeroing all fingers...");
        std::vector<float> zeroPositions(20, 0.0f);  // 5 fingers × 4 joints
        std::vector<float> zeroWristPositions(2, 0.0f);  // 2 wrist joints
        client.sendHandStreams(zeroPositions, 0.45f);
        client.sendWristStreams(zeroWristPositions);
        std::this_thread::sleep_for(std::chrono::seconds(1));

        // Test each joint of each finger
        // Joint layout: thumb[0-3], index[4-7], middle[8-11], ring[12-15], pinky[16-19]
        for (size_t fingerIdx = 0; fingerIdx < fingers.size(); ++fingerIdx)
        {
            for (int j = 0; j < 4; ++j)
            {
                int jointIdx = fingerIdx * 4 + j;
                utils::printSection(
                    fingers[fingerIdx] + " - " + jointNames[j] +
                    " (joint " + std::to_string(jointIdx) + ")");

                // Determine range based on joint and finger
                float minDeg = (j == 0 && fingers[fingerIdx] != "thumb") ? -30.0f : 0.0f;
                float maxDeg = (j == 0 && fingers[fingerIdx] != "thumb") ? 30.0f : 90.0f;

                float minRad = minDeg * M_PI / 180.0f;
                float maxRad = maxDeg * M_PI / 180.0f;

                // Run cycles (use streaming for high-frequency commands)
                for (int cycle = 0; cycle < cycles; ++cycle)
                {
                    // Start with all zero (20 positions: 5 fingers × 4 joints)
                    std::vector<float> positions(20, 0.0f);

                    // For distal joint (j==3), pre-flex intermediate
                    if (j == 3 && fingers[fingerIdx] != "thumb")
                    {
                        positions[jointIdx - 1] = 90.0f * M_PI / 180.0f; // intermediate joint
                    }

                    // Move to max position (use streaming for high-frequency control)
                    positions[jointIdx] = maxRad;
                    client.sendHandStreams(positions, 0.45f);
                    client.sendWristStreams(zeroWristPositions);
                    std::this_thread::sleep_for(
                        std::chrono::duration<double>(delay));

                    // Move to min position
                    positions[jointIdx] = minRad;
                    client.sendHandStreams(positions, 0.45f);
                    client.sendWristStreams(zeroWristPositions);
                    std::this_thread::sleep_for(
                        std::chrono::duration<double>(delay));
                }
            }
        }

        // Return to zero (use streaming mode)
        utils::printSection("Returning to zero position...");
        client.sendHandStreams(zeroPositions, 0.45f);
        client.sendWristStreams(zeroWristPositions);
        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        // Disable streaming mode
        utils::printSection("Disabling streaming mode...");
        client.setStreamingMode(false);
        std::this_thread::sleep_for(std::chrono::milliseconds(200));

        utils::printSuccess("Test hand demo completed!");
        return 0;
        
    } catch (const SdkException& e) {
        utils::printError(std::string("Demo failed: ") + e.what());
        std::cout << "\nMake sure prohand-headless-ipc-host is running\n";
        return 1;
    } catch (const std::exception& e) {
        utils::printError(std::string("Unexpected error: ") + e.what());
        return 1;
    }
}
